# Python learning tutorial Box 
### Author: xldang2020@gmail.com

* [1.Introduction](#intro)
* [2.Box artchitecture](#artchi)
* [3.Calculator](#calculator)
    * [3.1 Data type](#ch3.1)
    * [3.2 Operator](#ch3.2)
    * [3.3 Once for all: function](#ch3.3)
* [4.Vocabulary notebook](#vocab)
    * [4.1 list](#ch4.1)
    * [4.2 dict](#ch4.2)
    * [4.3 set](#ch4.3)
    * [4.4 Progamming a Vocabulary notebook app.](#ch4.4)
* [5.Searching toolkit](#search)
    * [5.1 import](#ch5.1)
    * [5.2 loop control](#ch5.2)
    * [5.3 Guess me?](#ch5.3)
* [6.OOP (object-oriented programming)](#OOP)
    * [6.1 Learning OOP with Python](#ch6.1)
    * [6.2 Calculator with OOP](#ch6.2)
* [7.Summary and outlook](#summary)

## 1. Introduction <a name='intro'></a>


**If you are new to programming, the material in this article serves as an introduction to the basic building blocks for writing code in Python language.** 

**This article follows this path to learn the basic knowledge of Python:Want programming a common application? Yes! Let's do it piece by piece.**

## 2.Box artchitecture <a name='artchi'></a>

<img src='./img/Python Freshman Toolkits.png' >

## 3.Calculator  <a name='calculator'></a>

A simple calculator is not simple. Let's try!

### 3.1 Data type / 数据类型 <a name='ch3.1'></a>

- int 整型:such as 1,2,3,100,200,...
- float 浮点型: such as 1.0,2.0,3.5,5.02,...
- bool 布尔型 ：True,False

### 3.2 Operator / 操作符  <a name='ch3.2'></a>
- +
- -
- \* (×)
- / (÷)
- // (整除)
- % (取模：返回除法运算后的余数)
- << (左移：二进制)
- \>> (右移：二进制)
- \> (大于) 、\>= (大于等于)、\< (小于) 、\<=(小于等于)、\== (等于)、\!= (不等于)
- not （布尔“非”）、and （布尔“与”）、or （布尔“或”）

### 3.3 Once for all: function  / 一劳永逸：函数  <a name='ch3.3'></a>

***def*** is the head of *definition*.  [ˌdefɪˈnɪʃn] 定义

In Python language, ***def*** is used to define a function.

In [1]:
def add(a=0,b=0):
    x=a+b
    return x

In [2]:
add(1,2)

3

In [3]:
def minus(a=0,b=0):
    x=a-b
    return x

In [4]:
minus(1,2)

-1

In [5]:
def multiply(a=0,b=0):
    x=a*b
    return x

In [6]:
multiply(1,2)

2

### 3.4 My calculator app. : Freshman Version <a name='ch3.4'></a>

**What you need to know:**
- Expression: if (while,for, will be learn later)
- Function call: a function appears in another function's body.

Let's combine lots of opertors in one function.

In [7]:
def myCalculator(a=0,b=0,operator='+'):
    if operator=='+':
        return add(a,b)
    elif operator=='-':
        return minus(a,b)
    elif operator=='*':
        return multiply(a,b)
    else:
        print("Want more function? Try yourself!")

In [8]:
myCalculator(2,3,'*')

6

## 4. Vocabulary notebook <a name='vocab'></a>

**What you need to know:**

- list:列表
- dict: dictionary 字典
- set: 集合
- tuple: 元组

### 4.1 list 列表  <a name='ch4.1'></a>

In Python,if you define a variable *x* with [ something ],then *x* is a list object.

In [9]:
x=[1,2,3,'hello','a','b','c']

In [10]:
type(x)

list

- What is **in** mean?

In [11]:
1 in x

True

In [12]:
5 in x

False

As you see above, if *a*  contains *b* ,then *b in a* return True else return False

- **Select values from a list:**

In [13]:
x[0]

1

In [14]:
x[3]

'hello'

In [15]:
x[2:4]

[3, 'hello']

In [16]:
x[-1] #the last one

'c'

**built-in function：Python内置函数**

- len :测量一个对象的容量
- max/min :最大/最小

In [17]:
len(x)

7

In [18]:
max(x) # x中必须是可以相互比较的元素

TypeError: '>' not supported between instances of 'str' and 'int'

In [ ]:
max([1,2,3,4,5]),min(['a','aa','aaa'])

**初次接触：面向对象编程——类与对象**
- list is a python ***class***
- y=[1,2] is a list ***object***, and a object has many ***methods***.
- y.append(3) is a add_value method

In [19]:
y=[1,2]
y.append(3)
print(y)

[1, 2, 3]


Want more details? Let's talk about it later!

### 4.2 dict 字典  <a name='ch4.2'></a>

In Python,if you define a variable x with { key1:value1,key2:value2,... },then x is a list object.

In [20]:
dictX={'Name':'Lucy','Age':18}
type(dictX)

dict

In [21]:
# read values from dict in this way:
dictX['Name']

'Lucy'

In [22]:
# add values into dict in this way:
dictX['address']='American'
print(dictX)

{'Name': 'Lucy', 'Age': 18, 'address': 'American'}


In [23]:
# If a key is alread in a dict,what will happen?
dictX['address']='China'

In [24]:
#If a key is alread in a dict,it's value will be replaced!
print(dictX)

{'Name': 'Lucy', 'Age': 18, 'address': 'China'}


In [25]:
#  How can i know  if a key is in one dict object
'Age' in dictX

True

In [26]:
#  How can i know  if a key is in one dict object
'Gender' in dictX

False

### 4.3 set 集合  <a name='ch4.3'></a>

In Python,if you define a variable **x** with {value1,value2,... },then **x** is a ***set*** object.

In [27]:
setX={1,2,3,4,5,7}
type(setX)

set

集合（Set）是简单对象的无序集合（Collection）,这里的集合概念和数学上的集合概念是一样的：**无序、元素不重复**

In [28]:
setX={1,2,3,4,5,7}
print('Before add in:',setX)
setX.add(1) # add in a value which already exists in a set object, nothing changes
print('After add in:',setX,'nothing happens!')

Before add in: {1, 2, 3, 4, 5, 7}
After add in: {1, 2, 3, 4, 5, 7} nothing happens!


In [29]:
setX={1,2,3,4,5,7}
print('Before add in:',setX)
setX.add(8) # add in a new value , set changes
print('After add in:',setX,'new value join in!')

Before add in: {1, 2, 3, 4, 5, 7}
After add in: {1, 2, 3, 4, 5, 7, 8} new value join in!


In [30]:
setY={1,2,3,9}
print('setX is:',setX)
print('setY is:',setY)

ans=setX.intersection(setY) # assign to ans 
print("Their intersection(交集) is: ",ans)

setX is: {1, 2, 3, 4, 5, 7, 8}
setY is: {1, 2, 3, 9}
Their intersection(交集) is:  {1, 2, 3}


**第二次接触：面向对象编程——类与对象**
- set is also a python ***class***
- setY={1,2,3} is a set ***object***, and a ***object*** has many ***methods***.
- setY.intersection(setX) is a intersection(交集) method

Try another method:

In [31]:
setX.difference(setY) # 取setX与setY的差集

{4, 5, 7, 8}

In [32]:
setX.union(setY) # 取setX与setY的并集

{1, 2, 3, 4, 5, 7, 8, 9}

### 4.4 Progamming a Vocabulary notebook app. 做一个单词本  <a name='ch4.4'></a>

In [33]:
myVocab={} # 先定义一个空的字典
def useMyVocab(word='',means='',myNote=myVocab):
    if word not in myNote:
        myNote[word]=[means] # a list object as key's value
    elif means not in myNote[word]:
        myNote[word].append(means)
    else:
        print("Your notebook already has a word named %s with means %s"%(word,means))
    return myNote

In [34]:
useMyVocab('Apple','苹果',myVocab)

{'Apple': ['苹果']}

In [35]:
useMyVocab('Apple','苹果',myVocab)

Your notebook already has a word named Apple with means 苹果


{'Apple': ['苹果']}

In [36]:
useMyVocab('Apple','苹果手机',myVocab)

{'Apple': ['苹果', '苹果手机']}

In [37]:
useMyVocab('Orange','橙子',myVocab)

{'Apple': ['苹果', '苹果手机'], 'Orange': ['橙子']}

In [38]:
useMyVocab('Orange','橙色',myVocab)

{'Apple': ['苹果', '苹果手机'], 'Orange': ['橙子', '橙色']}

In [39]:
myVocab

{'Apple': ['苹果', '苹果手机'], 'Orange': ['橙子', '橙色']}

In [40]:
myVocab['Apple']

['苹果', '苹果手机']

## 5. Searching toolkit <a name='search'></a>

Suppose there is a number **X** belongs to [-10000,10000], how can you guess it right as soon as possible, and how many times do you need?

***What you need to know:***

- random : a Python package 随机数模块包
- while : 循环控制流
- if : 条件控制
- for :  循环控制流
- break: 停止循环控制流
- continue: 跳出当前循环



### 5.1 import  <a name='ch5.1'></a>

**第一次接触：开源编程——引入外部依赖包 import**

In [41]:
import random  # then random can be used as a object variable

In [42]:
random.randint(0,10)

0

In [43]:
random.uniform(2,5)

3.614438408032243

In [44]:
import numpy as np

In [45]:
np.pi # π

3.141592653589793

In [46]:
np.sin(np.pi/2)

1.0

In [47]:
np.abs(-2)

2

In [48]:
np.sqrt(4)

2.0

In [49]:
np.square(3)

9

### 5.2 Loop Control 循环控制流  <a name='ch5.2'></a>

**第一次接触：开源编程——引入外部依赖包 import**

### 5.2.1 While loop
- while flag:
    - event A
        
**while** body (event A) will never ends if **flag** is ***True*** unless **flag** turn into ***False***.

In [50]:
times=0
while times<10:
    #event A
    print("Flag times<10 is bool:",times<10)
    times=times+1
    print("  Event A goes %d times..."%times)
print("The last Flag times<10 is bool:",times<10)

Flag times<10 is bool: True
  Event A goes 1 times...
Flag times<10 is bool: True
  Event A goes 2 times...
Flag times<10 is bool: True
  Event A goes 3 times...
Flag times<10 is bool: True
  Event A goes 4 times...
Flag times<10 is bool: True
  Event A goes 5 times...
Flag times<10 is bool: True
  Event A goes 6 times...
Flag times<10 is bool: True
  Event A goes 7 times...
Flag times<10 is bool: True
  Event A goes 8 times...
Flag times<10 is bool: True
  Event A goes 9 times...
Flag times<10 is bool: True
  Event A goes 10 times...
The last Flag times<10 is bool: False


### 5.2.2 for loop

- for v in iteration_object:
    - event A (handle v)
    - if v satisfies one condition, we can ***break*** this **for** body
    
**① what is iteration_object?** 什么是可迭代对象

In [51]:
range(10)

range(0, 10)

In [52]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


**② 字符串，列表，元祖，字典，集合等等，都是可迭代对象。**

In [53]:
aList=['1','2',3,4,5,'a','b']
for v in aList:
    print(v)

1
2
3
4
5
a
b


In [54]:
# built-in function len 
len(aList)

7

In [55]:
#How to traverse a list by it's index ?如何按索引号遍历一个列表？
for index in range(len(aList)):
    print(aList[index])

1
2
3
4
5
a
b


### 5.2.3 if / break /continue 

**Let's do something in loop body**

**① 第一次接触：python standard input 标准输入流**

In [56]:
your_input=input("Please enter a integer value between (-10,10): ")

Please enter a integer value between (-10,10): 5


In [57]:
your_input

'5'

In [58]:
# it become a string object, 
type(your_input) 

str

In [59]:
# we need to force type changes like this :
int(your_input)

5

**② let's break something!**

In [60]:
while True:
    v=int(input("Please enter a integer value between (-10,10): "))
    if v==5:
        break
    else:
        print('Try again!')

Please enter a integer value between (-10,10): 6
Try again!
Please enter a integer value between (-10,10): 5


**③ let's continue!**

if you enter a wrong number,how can it continue?

In [61]:
while True:
    v=int(input("Please enter a integer value between (-10,10): "))
    if v>10 or v<-10:
        print('You enter a wrong number! Please try again!')
        continue
    elif v==5:
        break
    else:
        print('Try again!')

Please enter a integer value between (-10,10): 6
Try again!
Please enter a integer value between (-10,10): 12
You enter a wrong number! Please try again!
Please enter a integer value between (-10,10): 5


**④ for loop continue**

In [62]:
alist=[1,1,1,1,2,3,4,5]
print('Lets print non-1 values in alist!')
for v in alist:
    if v==1:
        continue
    print(v)

Lets print non-1 values in alist!
2
3
4
5


### 5.3 Guess me? <a name='ch5.3'></a>

**Let's back to our mission:**
- Suppose there is a number X belongs to [-10000,10000],provide by ***random*** package;
- How can you guess it right as soon as possible;
- How many times do you need?

In [63]:
def guessMe():
    from random import randint
    target=randint(-10000,10000) # includes -10000,10000
    times=0
    smaller,bigger=-10000,10000
    ## check if illegal input!
    while True:
        x=int(input("Please guess me a number between(%d~%d): "%(smaller,bigger)))
        if x>bigger or x<smaller:
            continue
        else:
            break
    ## guess mission main body
    while True:
        if x==target:
            print("You guess Right! Costs %d"%times)
            break
        elif x>target:
            bigger=x-1
            times+=1
            x=int(input("You're Wrong! Please try again between(%d,%d): "%(smaller,bigger)))
        else:
            smaller=x+1
            times+=1
            x=int(input("You're Wrong! Please try again between(%d,%d): "%(smaller,bigger)))

In [64]:
guessMe()

Please guess me a number between(-10000~10000): 0
You're Wrong! Please try again between(-10000,-1): -5000
You're Wrong! Please try again between(-4999,-1): -2500
You're Wrong! Please try again between(-4999,-2501): -1000
You're Wrong! Please try again between(-4999,-1001): -3000
You're Wrong! Please try again between(-4999,-3001): -3500
You're Wrong! Please try again between(-4999,-3501): -4000
You're Wrong! Please try again between(-3999,-3501): -3800
You're Wrong! Please try again between(-3799,-3501): -3700
You're Wrong! Please try again between(-3799,-3701): -3750
You're Wrong! Please try again between(-3799,-3751): -3770
You're Wrong! Please try again between(-3769,-3751): -3760
You're Wrong! Please try again between(-3759,-3751): -3755
You're Wrong! Please try again between(-3754,-3751): -3752
You're Wrong! Please try again between(-3754,-3753): -3753
You're Wrong! Please try again between(-3754,-3754): -3754
You guess Right! Costs 15


***There is a bug in former guessMe function, do you notice it?***

You're Wrong! Please try again between(-4999,-1001): -3000  
You're Wrong! Please try again between(-4999,-3001): -3500  
You're Wrong! Please try again between(-4999,-3501): -4000  

**It does not check out input value's constraint！**  
**Let's fix it!**

In [65]:
# let's program a help function do value-check job!
def legalInput(smaller,bigger):
    while True:
        x=int(input("Please guess me a number between(%d~%d): "%(smaller,bigger)))
        if x>bigger or x<smaller:
            print("Careful your input! Input should belongs to(%d~%d): "%(smaller,bigger))
            continue
        else:
            break
    return x

In [67]:
legalInput(-10,10)

Please guess me a number between(-10~10): 12
Careful your input! Input should belongs to(-10~10): 
Please guess me a number between(-10~10): 5


5

In [68]:
def guessMeRight():
    from random import randint
    target=randint(-10000,10000) # includes -10000,10000
    times=0
    smaller,bigger=-10000,10000
    ## check if illegal input!
    x=legalInput(smaller,bigger)
    ## guess mission main body
    while True:
        if x==target:
            print("You guess Right! Costs %d times."%times)
            break
        elif x>target:
            bigger=x-1
            times+=1
            x=legalInput(smaller,bigger)
        else:
            smaller=x+1
            times+=1
            x=legalInput(smaller,bigger)

In [69]:
guessMeRight()

Please guess me a number between(-10000~10000): 0
Please guess me a number between(1~10000): 5000
Please guess me a number between(1~4999): 3000
Please guess me a number between(1~2999): 1500
Please guess me a number between(1501~2999): 3000
Careful your input! Input should belongs to(1501~2999): 
Please guess me a number between(1501~2999): 2000
Please guess me a number between(1501~1999): 1800
Please guess me a number between(1501~1799): 1600
Please guess me a number between(1501~1599): 1550
Please guess me a number between(1551~1599): 1570
Please guess me a number between(1571~1599): 1580
Please guess me a number between(1571~1579): 1575
Please guess me a number between(1571~1574): 1572
You guess Right! Costs 11 times.


## 6. OOP (object-oriented programming) <a name='OOP'></a>

**Object-oriented programming (OOP)** is a programming paradigm based on the concept of "objects", which can contain data and code: data in the form of fields (often known as attributes or properties),  and code, in the form of procedures (often known as methods).  

高级编程语言主要分为：编程面向对象编程和面向过程编程。
- **面向过程（Procedure Oriented 简称PO ：如C语言）：**

    - 从名字可以看出它是注重过程的。当解决一个问题的时候，面向过程会把事情拆分成： 一个个函数和数据（用于方法的参数） 。然后按照一定的顺序，执行完这些方法（每个方法看作一个过程），等方法执行完了，事情就搞定了。

- **面向对象（Object Oriented简称OO ：如C++，JAVA等语言）：**

    - 看名字它是注重对象的。当解决一个问题的时候，面向对象会把事物抽象成对象的概念，就是说这个问题里面有哪些对象，然后给对象赋一些属性和方法，然后让每个对象去执行自己的方法，问题得到解决。

**用1个例子说明两者的区别：**[@知乎·呜呜轩轩](https://zhuanlan.zhihu.com/p/75265007)


问题： 洗衣机里面放有脏衣服，怎么洗干净？

面向过程的解决方法：

1、执行加洗衣粉方法；

2、执行加水方法；

3、执行洗衣服方法；

4、执行清洗方法；

5、 执行烘干方法；

以上就是将解决这个问题的过程拆成一个个方法（是没有对象去调用的），通过一个个方法的执行来解决问题。

面向对象的解决方法：

1、我先弄出两个对象：“洗衣机”对象和“人”对象

2、针对对象“洗衣机”加入一些属性和方法：“洗衣服方法”“清洗方法”、“烘干方法”

3、针对对象“人”加入属性和方法：“加洗衣粉方法”、“加水方法”

4、然后执行

人.加洗衣粉

人.加水

洗衣机.洗衣服

洗衣机.清洗

洗衣机.烘干

解决同一个问题 ，面向对象编程就是先抽象出对象，然后用对象执行方法的方式解决问题。



### 6.1 Learning OOP with Python <a name='ch6.1'></a>

**What you need to know:**  
**①面向对象编程的特性**
- 封装，
- 继承，
- 多态.

**②class and object(类与对象)、attributts and methods(属性和方法)**  
- **Class attributes(类属性)** – belong to the class as a whole; there is only one copy of each one
- **Instance attributes(实例或对象属性)** – data that belongs to individual objects; every object has its own copy of each one
- **Instance methods(实例或对象方法)** – belong to individual objects, and have access to instance variables for the specific object they are called on, inputs, and class variables
- **Advanced:**
    - **Class methods(类方法)** – belong to the class as a whole and have access only to class variables and inputs from the procedure call
**Let's figure out what are those means!**  


### 6.1.1 封装 

封装，就是把客观事物封装成抽象的类(class)，并且类可以把自己的数据和方法只让可信的类或者对象操作

**Imagine a thing a life...such as Dogs**

In [70]:
class Dogs():
    count=0
    def __init__(self,name='',age=0):
        self.name=name
        self.age=age
        Dogs.count+=1
    def whichDog(self):
        print("This is Dog %s, and %d years old."%(self.name,self.age))
    def jobDoing(self,job=''):
        print("Dog %s usually doing %s with human!"%(self.name,job))

In [71]:
alaska=Dogs('alaska',3)

In [72]:
alaska.whichDog()

This is Dog alaska, and 3 years old.


In [73]:
alaska.jobDoing('fun')

Dog alaska usually doing fun with human!


In [74]:
shepherd=Dogs('shepherd',4)

In [75]:
shepherd.whichDog()

This is Dog shepherd, and 4 years old.


In [76]:
shepherd.jobDoing('protection')

Dog shepherd usually doing protection with human!


In [77]:
print("How many object-dogs are created by class-Dogs? \nThe answer is :%d "%Dogs.count)

How many object-dogs are created by class-Dogs? 
The answer is :2 


**What you should care about in upper class are:**

- self,用来把参数传递给类class的对象
- \__init__, 用来初始化一个对象的属性: self.name 和self.age
- 如果你有一个没有参数的方法methods，你依旧必须拥有一个参数—— self ,如 whichDog;
- 就上述Dogs类而言：
    - count 是类属性,所有对象共享；
    - self.name和self.age 是对象属性；
    - whichDog和barking 是对象方法

### 6.1.2 继承 Inheritance 与多态 Polymorphism

就和字面意思差不多，继承是一个类对另一个类的代码重用。  
面向对象编程的一大优点是对代码的重用（Reuse），重用的一种实现方法就是通过继承（Inheritance）机制。

**假如随着动物的进化，狗中贵族哈士奇拥有了新的技能：说人话！！！ 这时候需要为之重新创造一个新狗类吗？**  
答案是：不需要！  
需要做的是：
- 继承：哈士奇虽然会说了人话，但它还是狗！还具有原来的属性；
- 多态：毕竟会说了人话，在jobDoing介绍的时候自然要炫耀一下：重写对象方法；

In [78]:
class speakHusky(Dogs):
    def __init__(self,name,age,speakLanguage='English'):
        Dogs.__init__(self,name,age)
        #super().__init__(self,name,age)
        self.speak=speakLanguage
    def jobDoing(self,job=''):
        print("Dog %s usually doing %s, And talk %s with master!"%(self.name,job,self.speak))
    def halfHuman(self):
        print("Dog %s can be called half-human now!"%self.name)

In [79]:
husky=speakHusky('Husky',5,'English')

In [80]:
husky.whichDog()

This is Dog Husky, and 5 years old.


In [81]:
husky.jobDoing('fun')

Dog Husky usually doing fun, And talk English with master!


In [82]:
husky.halfHuman()

Dog Husky can be called half-human now!


**Traceback...**
- 父类属性 self.name,self.age 被子类speakHusky完全继承；
- 父类方法 whichDog 被子类speakHusky完全继承；
- 父类方法 jobDoing 被子类speakHusky重写（多态）；
- 子类speakHusky 进化出了新技能 halfHuman。

### 6.2 Calculator with OOP<a name='ch6.2'></a>

**学习了类与对象的知识，你就拥有了强大的技能，因为万物皆可对象！**  
**Now you say "Python OOP makes me strong as mountains!"**

**So rebuilding Calculator with OOP is now a simple matter.**

[callback:Calculator-Freshman Version](#ch3.4)

**① 基本计算器**

In [83]:
class BasicCalculator():
    def __init__(self,name=''):
        self.name=name
    def info(self):
        print("This is %s Calculator!"%self.name)
    def plus(self,x=0,y=0):
        return x+y
    def minus(self,x=0,y=0):
        return x-y
    def multiply(self,x=0,y=0):
        return x*y
    def division(self,x=0,y=1):
        if y==0 or y==0.0:
            print("Wrong input: Division by Zero!")
            return
        else:
            return x/y

In [84]:
basic=BasicCalculator('Basic')

In [85]:
basic.info()

This is Basic Calculator!


In [86]:
basic.plus(1,2)

3

In [87]:
basic.minus(3,4)

-1

In [88]:
basic.multiply(3,4)

12

In [89]:
basic.division(2,0)

Wrong input: Division by Zero!


In [90]:
basic.division(6,3)

2.0

**②科学计算器**

In [91]:
import math

class sciCalculator(BasicCalculator):
    def __init__(self,name=''):
        BasicCalculator.__init__(self,name)    
    def plus(self,x,y):
        # 不限制数据类型
        return x+y
    def sin(self,x=0):
        return math.sin(x)
    def cos(self,x=0):
        return math.cos(x)
    def tan(self,x=0):
        return math.tan(x)

In [92]:
sci=sciCalculator('Scientific')

In [93]:
sci.info()  # 完全继承

This is Scientific Calculator!


In [94]:
sci.plus('a','b') # 重写、多态

'ab'

In [95]:
sci.sin(math.pi/6) # 新对象方法

0.49999999999999994

In [96]:
sci.cos(math.pi/3) # 新对象方法

0.5000000000000001

In [97]:
sci.tan(math.pi/4)# 新对象方法

0.9999999999999999

## 7. Summary and outlook <a name='summary'></a>

### Summary

**For the Python programming language, the content of this tutorial is very basic and far from comprehensive, but it can be used as an amazing date to discover the simplicity and fun of Python.**

### Outlook

- to newbies, this content maybe as a quick glance;
- to beginners, I sugguest this book [《A Byte of Python-简明 Python 教程 v.4.07c 2017》 提取码：y4sn](https://pan.baidu.com/s/1pwaZOAl7GVjX5zs03mXA9A);
- to senior, I look forward to your feedback and guidance.[Feedback](http://fallincloud.com/?p=239)

***Sincerely,I hope this tutorial can be helpful to you guys who learn python programming***

<h5 style="text-align:right">© Copyright 2021-, xldang's Python classroom.</h5>

<h5 style="text-align:right"> Email:xldang2020@gmail.com </h5>

<h5 style="text-align:right"> Last updated on Apr 10, 2021. </h5>